nasa/ow_simulator
# Oceanwater 945 ArmMoveCartesianGuarded added

## Linked Issues:
| EPIC ⚡ | [OCEANWATER-933](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-933) |
| :----------- | :----------- |
| Jira Ticket 🎟️ | [OCEANWATER-945](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-945) |
| Github :octocat:  | # |

## Summary of Changes
* Added ArmMoveCartesianGuarded
* There is now a mixin for arm actions that modify pose. 
* Reduced code repetition where possible between ArmMoveCartesian and actions not yet added (this will likely be on-going). 

## Reviewer Assignments
@anthonykoutroulis To save your own time, please only perform a review of the code.
@kmdalal Perform the following tests, and review the code at your discretion.

## Test Help Message
Test help message and proof-read
```
$ rosrun ow_lander arm_move_cartesian_guarded.py --help  
usage: arm_move_cartesian_guarded.py [-h] [--frame {0,1}] [--relative] [-x X] [-y Y] [-z Z]
                                     [--euler X Y Z | --quaternion X Y Z W] [--force FORCE]
                                     [--torque TORQUE]

Move end-effector by Cartesian coordinates/translation.

optional arguments:
  -h, --help            show this help message and exit
  --frame {0,1}, -f {0,1}
                        The frame index corresponding to the frame that orientation and position are
                        relative to. {0: 'base_link', 1: 'l_scoop_tip'} (default: 0)
  --relative, -r        If True, pose will be interpreted as relative to the current pose. (default: False)
  -x X                  X-coordinate or the change in the X-coordinate (default: 0)
  -y Y                  Y-coordinate or the change in the Y-coordinate (default: 0)
  -z Z                  Z-coordinate or the change in the Z-coordinate (default: 0)
  --euler X Y Z, -e X Y Z
                        Orientation in Euler angle form. Takes 3 radian values. (default: [0, 0, 0])
  --quaternion X Y Z W, -q X Y Z W
                        Orientation in quaternion form. Takes 4 scalars, the first 3 are the imaginary
                        parts, and the 4th is the real part. (default: [0, 0, 0, 1])
  --force FORCE         Arm will stop when F/T sensor encounters a force above this value in Newtons
                        (default: 200)
  --torque TORQUE       Arm will stop when F/T sensor encounters a torque above this value in Newton*meters
                        (default: 100)
```

## Test Action
1. Start simulation in atacama_y1a so that terrain is consistent across tests.
2. Prime scoop to poke ground with 
`rosrun ow_lander arm_move_cartesian.py  -x 1.7 -e 3.14 0 0`. 
The scoop should now be just above the ground.
3. Perform a guarded relative move of the scoop down and into the ground with 
`rosrun ow_lander arm_move_cartesian_guarded.py -r -z 0.6` 
The following output should result. Force value will vary, but should be above 200 N.
```
ArmMoveCartesianGuarded: Succeded - ArmMoveCartesianGuarded trajectory stopped by a force of 309.77 N
```
4. Lift scoop off the ground and up high
`rosrun ow_lander arm_move_cartesian.py  -x 1.7 -z 0.5 -e 3.14 0 0`
5. Call the same relative guarded move command as before 
`rosrun ow_lander arm_move_cartesian_guarded.py -r -z 0.6` 
Scoop should stop just above the ground. The following output should result
```
ArmMoveCartesianGuarded: Succeded - ArmMoveCartesianGuarded trajectory completed without breaching force or torque thresholds
```
6. Call the same relative guarded move command but now increasing the force tolerance to something unreachable
`rosrun ow_lander arm_move_cartesian_guarded.py -r -z 0.6 --force 600` 
Scoop will hit the terrain and the following error output result
```
ArmMoveCartesianGuarded: Aborted - Arm failed to reach pose despite neither force nor torque thresholds being breached. Likely the thresholds were set too high or a planning error occurred. Try a lower threshold.
```
7. While scoop is still planted on the ground, cause a torque and force threshold breach by sliding it horizontally and setting a low torque threshold.
`rosrun ow_lander arm_move_cartesian_guarded.py -r -y 0.1 --torque 10`
Note that force threshold is should now be breached because we used the default of 200 N. Output should be
```
ArmMoveCartesianGuarded: Succeded - ArmMoveCartesianGuarded trajectory stopped by a force of 368.85 N and a torque of 31.97 Nm
```
8. Finally try and get only a torque threshold breach by doing the same maneuver but making the force threshold unreachable. 
`rosrun ow_lander arm_move_cartesian_guarded.py -r -y 0.1 --force 600 --torque 10`
```
ArmMoveCartesianGuarded: Succeded - ArmMoveCartesianGuarded trajectory stopped by a torque of 56.07 Nm
```
